docs/design/2022-11-17-ttl-table.md

```diff
@@ -0,0 +1,357 @@
+
+#### Time Zone Consideration
+
+Currently, we support three field types as the time column for TTL: `Date`, `DateTime` and `TimeStamp` . However, for `Date` and `DateTime`, they are not an absolute time, and we need another "time zone" information to determine what their accurate time point is. We have two options to select the time zone:
+
+1. Use the system time zone which can be fetched from global variable `system_time_zone` and is determined when the cluster bootstraps. Because the `system_time_zone` will never changes, the row will be deleted at a certain time. However, it will have some problem when the user sets the cluster time zone `time_zone` to a different value other than `SYSTEM`. For example, if the `system_time_zone` is `+08:00` and `time_zone` is set to `UTC`, the record will be seen as expired immediately when the `tidb_ttl_job_run_interval` is less than 8 hours.
+
+2. Use the cluster time zone `time_zone`. It is fine when cluster time zone `time_zone` does not change. But if user changes the `time_zone` settings, the TTL job should be aware of it in time to forbid deleting some unexpected rows. In this case, whether one row is expired or not can not be determined without knowing the `time_zone` at that time.
+
+### New System Variables
+
+Some new system variables will be introduced:
+
+- `tidb_ttl_job_enable`
+  - When this variable is `OFF`, the cluster will stop to schedule TTL jobs and the running jobs will be cancelled.
+  - Scope: Global
+  - Values: [ON, OFF]
+  - Default: ON
+
+- `tidb_ttl_job_run_interval`
+  - The schedule interval between two jobs for one TTL table
+  - Scope: Global
+  - Range: [10m0s, 8760h0m0s]
+  - Default: 1h
+
+- `tidb_ttl_job_schedule_window_start_time`
+  - This variable is used to restrict the start time of the time window of scheduling the ttl jobs. 
+  - Scope: Global
+  - Type: Time
+  - Default: 00:00 +0000
+
+- `tidb_ttl_job_schedule_window_end_time`
+  - This variable is used to restrict the end time of the time window of scheduling the ttl jobs.
+  - Scope: Global
+  - Type: Time
+  - Default: 23:59 +0000
+
+- `tidb_ttl_scan_worker_count`
+  - The count of the scan workers in each TiDB node
+  - Scope: Global
+  - Range: [1, 1024]
+  - Default: 4
+
+- `tidb_ttl_scan_batch_size`
+  - The limit value of each SELECT query in scan task
+  - Scope: Global
+  - Range: [1, 10240]
+  - Default: 500
+
+- `tidb_ttl_delete_worker_count`
+  - The count of the delete workers in each TiDB node
+  - Scope: Global
+  - Range: [1, 1024]
+  - Default: 4
+
+- `tidb_ttl_delete_batch_size`
+  - The batch size in one delete query when deleting expired rows.
+  - Scope: Global
+  - Range: [0, 10240]
+  - Default: 100
+
+- `tidb_ttl_delete_rate_limit`
+  - The rate limit of the delete operations in each TiDB node. 0 is for no limit
+  - Scope: Global
+  - Range: [0, MaxInt64]
+  - Default: 0
+
+- `tidb_ttl_enable_instance_worker`
```

Comment:
I can see users getting confused between `tidb_ttl_job_enable` and this. Whenever there are multiple things which need to be set, it can create a usability problem. Arguably there is also a third option to temporarily disable the TTL on the table level too :( 

A suggestion to reduce the configuration - remove (`tidb_ttl_job_enable`)? If it is required to terminate all running jobs, could it be SQL-like syntax instead?

Semi-related: I would also like there to be SQL syntax to synchronously block until TTL jobs are complete. i.e. something like:
```sql
ALTER TABLE t1 OPTIMIZE TTL;
```

Reply:
`tidb_ttl_enable_instance_worker` is used to forbid one tidb node running TTL jobs. By using the variable we can force TTL jobs and TP SQLs to running in different tidb. It is true that it may bring some confuse, maybe we can rename it or move it to the configurations or just remove it for the first version?

`tidb_ttl_job_enable` means to disable TTL job scheduling and it is some what different with a command to "stop" all running jobs. `tidb_ttl_job_enable` that means it not only stops the current running jobs but also means the job is not reschedule in the future unless it is opened again.

Reply:
For: `ALTER TABLE t1 OPTIMIZE TTL;` do you mean to trigger a TTL job manually? I think we can use admin statement like `ADMIN OPTIMIZE TTL <table_name>` , `ALTER TABLE ...` seems a DDL, but it is not.

Reply:
> Is it possible to rely on users disabling the TTL on the table-level only

Yes, it is supported. TTL table has an option `TTL_ENABLE=ON/OFF` , the default value is ON. When it is set to OFF, that means a table's TTL job is disabled.

> For ALTER TABLE t1 OPTIMIZE TTL; ...

Maybe we can introduce a new command to trigger a TTL job manually. If we do not want to use super privilege to tirgger a job, stopping a ttl job should have a same problem.

Reply:
Personally, I don't like `tidb_ttl_enable_instance_worker`, this variable was actually introduced because our TTL task scheduling between tidb nodes is not balanced enough, so we need to bring the management burden to the user.

I prefer removing it and improving the TTL task scheduling in the future.

@lcwangchao

Reply:
@morgo 

for`ALTER TABLE t1 OPTIMIZE TTL`, I feel it is not necessary to have it.
TTL tasks can be controlled by `tidb_ttl_job_run_interval`, you can change it and then observe the metrics or `tidb_ttl_table_status`

Reply:
The concern I have is with multiple settings to disable the TTL workers, garbage might accumulate. When debugging, you might want to ensure the background processes are up to date. But it can always be added later as future work, so we don't have to agree on anything now.